### ISSUE TITLE
Conventions for links to specific tabs in "nested" tabs like the ControlPanel

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#3057]
I believe we need to arrive at standard conventions for modeling a nested tab interface such as the ControlCenter whereby it is eventually easy to do:

> `<<show-tab "$:/core/ui/ControlPanel" "Info/Advanced">>`

Or as shortcut for the ControlCenter only:

> `<<TWCP Info/Advanced>>`

Whereas `Info/Advanced` can be any tab in the hierarchy.

To achieve this, I think we need to rename and restructure some of the ControlPanel tiddlers (under the hood), so that a generic nesting can be infered from the title alone and respective state rules applied.

Right now, these titles for example are problematic:

* **$:/core/ui/ControlPanel/Advanced**
* **$:/core/ui/ControlPanel/TiddlerFields**

They should really be:

* **$:/core/ui/ControlPanel/Info/Advanced**
* **$:/core/ui/ControlPanel/Info/Advanced/TiddlerFields**

Otherwise we will have to invent a lot of bloat-fields only to define specifc state tiddlers — which, unfortunately, hardly any of the ControlPanel tabs do whereas the states should follow consistent rules. However, this seems quite necessary to eventually, reliably invoke something like:

> `<<show-tab "$:/core/ui/ControlPanel" "Info/Advanced/TiddlerFields">>`

If it wasn't the above but the below:

> `<<show-tab "$:/core/ui/ControlPanel" "TiddlerFields">>`

...how could that mechanism figure out that **TiddlerFields** is really a tab under **Advanced** which itself is under **Info** and also when to stop?

The only way I could think of is to invent a new field `tab-state` at the tiddler defining the tab that specifies the state tiddler used for the **tabs** macro... and then a `tab-parent` for each tiddler defining the relationship to the parent tab. However, I would really want to avoid any of this and simply infer from naming conventions. On the other hand, in a non-ControlCenter scenario, it seems we will need some `tab-state` + `tab-parent` mechanism to reuse the same pattern and then say:

> `<<show-tab "outermost parent" "some tab nested anywhere in the tab hierarchy of outermost parent">>`

Possibly, to cater for "polymorphism" in the sense of a tab possibly being contained in various "outermost parent"s, the above requires a field prefix, like:

> `<<show-tab "tab-foo" "outermost parent" "some tab nested anywhere in the tab hierarchy of outermost parent">>`

...which translates to:

* for the "outermost parent" tab container
* find "some tab nested anywhere in the tab hierarchy of outermost parent" and set the states accordingly
* using these fields `tab-foo-state` and `tab-foo-parent`

**Comment by pmario** [comment]
I like the idea, to make the TW "state-system" more human predictable, and give supporters the possibility to post "links" or "bundles" that let users activate a certain TiddyWiki state

**Comment by pmario** [comment]
There is a reason why "$:/ControlPanel" is named that way and not "$:/core/ui/ControlPanel" ... _and_ there has been some discussions about the name and `$:/` prefix or not ... somewhere around Nov 2013 ~alpha 11/12/13/14 

So changing it, IMO doesn't work out. `$:/ControlPanel` is one of the view "user visible" system tiddlers. 

All of them have special names, outside of the convention used otherwise. ` $:/AdvancedSearch` is an other one. ...

**Comment by pmario** [comment]
Here's how the `<<tabs>>` state system works at the moment. From a "state implementation" point of view it's not so deeply nested as for humans. 

eg: [Info: Advanced: Tiddler Fields](https://wikilabs.github.io/editions/link-to-tabs/#%24%3A%2Fcore%2Fui%2FControlPanel%2FTiddlerFields) is a "3rd level content"

BUT the state mechanism only uses 2 transclusion levels, to create a state variable. See explanation below. 

----------

The `<<tabs>>` macro supports these params: `tabsList, default, state, class, template, buttonTemplate`

 - "ControlPanel" only uses the first 2 so 
   - state [defaults to](https://tiddlywiki.com/#tabs%20Macro): `$:/state/tab`
 - tabs macro uses the [qualify-macro](https://tiddlywiki.com/#qualify%20Macro:%5B%5Bqualify%20Macro%20(Examples)%5D%5D%20%5B%5Bqualify%20Macro%5D%5D) to extend the state variable
 - qualify [is implemented](https://tiddlywiki.com/#%24%3A%2Fcore%2Fmodules%2Fmacros%2Fqualify.js), as secribed in [tw state mechanism](https://tiddlywiki.com/#StateMechanism)

So if we paste this code int the different ControlPanel tiddlers, we can see, how the states, for the containing `<<tabs>>` macros are created: 

```
<code><<qualify "$:/state/tab">></code><br>
<code><<transclusion>></code>
```

So if we copy it to: [Info: Advanced: Tiddler Fields](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FControlPanel%2FTiddlerFields)

it will give us: 

```
$:/state/tab--645621743
{$:/core/ui/ControlPanel/TiddlerFields|$:/core/ui/ControlPanel/TiddlerFields|||}
```

If we open the [ControlPanel: Info: Advanced: TiddlerFields](https://tiddlywiki.com/#%24%3A%2FControlPanel)

we will get this result

```
$:/state/tab-1808262310
{$:/ControlPanel|$:/core/ui/ControlPanel/TiddlerFields|||}
```

The cool think is, that these state variables are independent to the language uses! ... because they only wrok with core tiddler names. ... As long as they don't change the state variables won't change!

The state tiddler form above would be used, if we would add an other `tabs macro` into this tiddler. ... and so on

So even if it is 3rd level content, the qualify macro only uses 2 elements. ControlPanel and the tiddler itself. 

This implementation has some use-case specific advantages, that we would need to change, if we would implement your proposal 1:1

just some facts and thoughts,

**Comment by pmario** [comment]
The format, that the qualify macro uses is: https://tiddlywiki.com/#transclusion%20Variable:%5B%5BCore%20Variables%5D%5D%20%5B%5Btransclusion%20Variable%5D%5D

**Comment by Jermolene** [comment]
@tobibeer I do like the idea of a tab hierarchy based on hierarchical tiddler titles, and I'd encourage you to continue to explore it.

FYI when I have to solve this kind of problem in my own projects I use the advanced search to list the most recently modified system tiddlers, and then click one the tabs I want to set. I can then copy and paste the state tiddler titles from the search results.

**Comment by tobibeer** [comment]
> So changing it, IMO doesn't work out. $:/ControlPanel is one of the few "user visible" system tiddlers.

Actually, I never suggested to change the ControlPanel itself, but only the tabs in it.

**Comment by tobibeer** [comment]
@pmario: My concerns about inconsistent naming stemmed from the fact that I cannot leverage a system path prefix as an explicit state. So, the fact that most tabs macros in the controlpanel don't even define an explicit state is more or less problem NUMBER 1 that needs solving. But it only makes sense to solve, if another mechanism provides a way to inspect the "tab-nesting", either via naming conventions or via those additional fields. The latter probably are the more reliable of the two conventions.

**Comment by tobibeer** [comment]
To make it more clear: If I wanted to address a certain tab, I would set ALL state tiddlers that have the state specified for the particular tags macro. So, qualify become irrelevant. I simply set all existing states to the same tab, no matter where they are displayed / listed. In other words, I would switch all instances at the same time... for the same tab nesting that is.

However, I now realise this poses a problem! Why? Because if I never accessed the tab sequence before there are no state tiddlers for me to update just yet, w/ or w/o a unique state prefix. So, it looks like I'm back at square 1.

**Comment by tobibeer** [comment]
So, the real challenge then becomes that I need a way to emulate the qualify macro so as to arrive at the very tab I want to address, no matter what its name is or that of its "parent".

**Comment by tobibeer** [comment]
I suppose the solution is quite different after all...

1) We need a **$state** instead of / in addition to a **state** param for the tabs macro so as to allow specifying a **non-qualified** state! 

Who cares if I change the tabs in "THE ControlPanel" or all instances of it? Nobody, really. But right now, there is no way to have them all behave in unison.

2) Then we implement the **tab-state** (being the **$state**) and **tab-parent** pattern in addition and have a helper macro determine the parent chain in order to then set all state tiddlers as required, since now they are deterministic, independent from any location, e.g. this macro returns a bunch of action-setfield widgets depending on how deep the nesting goes to update all states in one go.

**Comment by pmario** [comment]
>Actually, I never suggested to change the ControlPanel itself, but only the tabs in it.

Ah, OK .. I did misinterprete this: `<<show-tab "$:/core/ui/ControlPanel" "Info/Advanced">>`

**Comment by pmario** [comment]
>I do like the idea of a tab hierarchy based on hierarchical tiddler titles, and I'd encourage you to continue to explore it.

I do like this idea too, but we can't use the caption or similar things, since they are translateable. ... As a supporter / developer I want to use the same state tiddler for every language. ...

**Comment by pmario** [comment]
>My concerns about inconsistent naming stem from the fact that I cannot leverage a system path prefix as an explicit state. 

That's true. ... ControlTiddler names evolved over time. Some tiddlers have been moved... and so on. 
That's why some of them are inconsistent. 

Finding good system tiddler names is really hard. ... So it would be good for new users, if we could give them a good advice for their own tiddlers.

**Comment by tobibeer** [comment]
I think the intermediate solution is even simpler than a **$state** variable and I'll be making a somewhat bold PR for it, mildly challenging the status quo on tab states.

**Comment by pmario** [comment]
... but I still would like to have a mechanism, that allows an easier way to handle "qualify" created states. For me #3059 is "fighting the symptoms" and therefor a "temporary" solution.

**Comment by tobibeer** [comment]
Not at all fighting symptoms. There is literally no point to targeting qualify. It is too arbitrary to create a solution for. Therefore, the solution is to avoid using qualify altogether. Or rather, if you want to address specific tabs: never use qualify.

**Comment by pmario** [comment]
>Therefore, the solution is to avoid using qualify altogether. Or rather, if you want to address specific tabs: never use qualify.

Yea, but using hardcoded states is just a convention. If I want to address someone elses tabs, I need to have a look at the code of the tiddler, to see which state is used. ...

Your OP proposed a "set all nested states" at once ... That's not possible with the PR atm.

**Comment by tobibeer** [comment]
> Your OP proposed a "set all nested states" at once ... That's not possible with the PR atm.

True. that is why I removed the "fixes" bit in the PR, after all ;-)

So, the above is really the longterm vision while the PR provides the necessary groundworks.

**Comment by pmario** [comment]
I'm fine..
